initial proxy service implementation

engelmi opened the pull request:
This PR addresses #95 by 
- implementing the proxy binary called by the systemd unit to request a service on a remote node
- adding a public API for the agent which the proxy binary calls
   -  `CreateProxy`: Creates and tracks a proxy within the agent and emits the `ProxyNew`signal on the internal bus
   - `RemoveProxy`: Currently a dummy
- adding a `ProxyMonitor` which uses the existing `Monitor` implementation and calls the `Ready` method on the agents proxy object

sdunnagan commented on src/agent/agent.c:
```diff
@@ -990,6 +998,87 @@ static const sd_bus_vtable internal_agent_vtable[] = {
+                UNUSED sd_bus_message *m, UNUSED void *userdata, UNUSED sd_bus_error *ret_error) {
+        hirte_log_info("Received remove proxy request");
+
+        // todo...
```
How about in capitals, "TODO"?

sdunnagan commented on src/manager/node.c:
```diff
@@ -1308,3 +1355,40 @@ void node_unsubscribe(Node *node, Subscription *sub) {
+        }
+
+        if (!agent_request_start(req)) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
```
I suggest more specific error messages here, to assist with debugging.

sdunnagan reviewed (approved):
Nice work!  Splitting the proxy service support into multiple stages seems a good idea.

engelmi committed:
Add proxy binary and template

This binary (and the template callin git) connects to the local agent
to create the requrested proxy service.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

engelmi committed:
Implement proxy support in agent

This listens to connections on the local bus from the hirte-proxy
app and creates proxy objects which the hirte manager
will handle.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson commented on src/proxy/main.c:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        _cleanup_free_ char *token = NULL;
+        token = strtok_r(rest, "__", &rest);
```
This feels a bit weird to me. I would rather pass in the nodename and the service name as separate options and then combine it however is needed internally rather than encode some way to combine them into one in the CLI interface.

engelmi replied:
I agree. This would be clearer to the user of the proxy binary. 
The `<node>__<service>` pattern I took from #95 where we use the service like `service@nodename_servicename.service` and pass it to the proxy via `%i`. Not sure how we could split it in the serivce file.

alexlarsson replied:
Ah, I see. Yeah, might be hard to do that then...

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1010,6 +1018,81 @@ static const sd_bus_vtable internal_agent_vtable[] = {
+ ************************************************************************/
+static int agent_method_create_proxy(UNUSED sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        Agent *agent = (Agent *) userdata;
+        if (agent == NULL) {
```
This should never be able to happen. Not sure its really helpful to check it. If anything this should be an assert() to catch that something really weird is going on.

engelmi replied:
Removed.

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1010,6 +1018,81 @@ static const sd_bus_vtable internal_agent_vtable[] = {
+        const char *node_name = NULL;
+        const char *unit_name = NULL;
+        const char *mode = NULL;
+        int r = sd_bus_message_read(m, "sss", &node_name, &unit_name, &mode);
```
I wonder if we should really expose the mode option for proxies. It seems a bit unclear what it will do, for example "mode==replace" could modify/replace other requests to start a service that are queued, and I don't think we want that to happen from a proxy request.

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1010,6 +1018,81 @@ static const sd_bus_vtable internal_agent_vtable[] = {
+        hirte_log_info("Received remove proxy request");
+
+        /*
+         TODO: Currently empty, has to implement:
```
Do we need RemoveProxy? I'm not sure.
We should probably track the lifetime of the client that called AddProxy() and remove the proxy when it dies though.

alexlarsson replied:
Actually, we do need a RemoveProxy that is called from the manager when the started service later exits.But that should be on the peer-to-peer bus, not on the api bus.

engelmi replied:
What about the case where the requiring service doesn't need the service on the other node anymore, e.g. when it itself exits? Shouldn't we clean-up then as well?

engelmi replied:
I was wondering if we could keep the proxy binary alive and keep passing the unit state updates from the manager to the agent and proxy - and the proxy would use [sd_bus_notify](https://www.freedesktop.org/software/systemd/man/sd_notify.html) to notify the systemd service manager. Then we would be able to remove the proxy in both cases, I think. Maybe this is too simplified, though. It would also mean that more resources are required permanently to keep the proxy binary running, not only on starting it. 
What do you think? @alexlarsson

alexlarsson replied:
So, when I designed this initially I deliberately made the proxy service exit after successful start, because that way we won't have lots of poxy service processes around wasting resources.

However, you're right, and we probably also want to know when the proxy is not needed anymore. I think we can do this by setting `StopWhenUnneeded=yes`, and adding an `ExecStop=` command, which calls this RemoveProxy method. And then yes, we *do* need it.

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1492,6 +1575,69 @@ bool agent_start(Agent *agent) {
         }
 
+<<<<<<< HEAD
+=======
```
some leftover merge stuff here

engelmi replied:
I think this was due the recent merge from martins config PR. Resolved it.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_errorf(
+                                proxy->request_message, SD_BUS_ERROR_FAILED, "Proxy service failed to start");
+        }
```
Maybe, else { print a warning about weird status } ?

engelmi replied:
Added an error log.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+        free(proxy->node_name);
+        free(proxy->unit_name);
+        free(proxy->mode);
+        free(proxy->object_path);
```
new() called agent_ref(), so this should unref

engelmi replied:
Added the `agent_unref`.

alexlarsson commented on src/manager/monitor_proxy.c:
```diff
@@ -0,0 +1,262 @@
+        subscription->monitor = proxy_monitor_ref(monitor);
+        subscription->free_monitor = (free_func_t) proxy_monitor_unref;
+        LIST_INIT(subscriptions, subscription);
+        LIST_INIT(all_subscriptions, subscription);
```
A bunch of this should be shared in a common function rather than duplicated like this.

engelmi replied:
Moved to `subscription_new` and renamed the existing functions to `create_<monitor>_subscription`.

alexlarsson commented on src/manager/monitor_proxy.c:
```diff
@@ -0,0 +1,262 @@
+
+        _cleanup_sd_bus_error_ sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_sd_bus_message_ sd_bus_message *message = NULL;
+        int r = sd_bus_call_method(
```
This should be sd_bus_call_method_async. We don't want to block the main thread of the manager waiting for the response.

engelmi replied:
Changed.

alexlarsson commented on src/manager/node.c:
```diff
@@ -377,6 +378,55 @@ static int node_match_heartbeat(UNUSED sd_bus_message *m, UNUSED void *userdata,
+        if (providing_node == NULL) {
+                hirte_log_errorf("Could not find node with name '%s'", node_name);
+                return -ENXIO;
+        }
```
It should also check that the target node is online.

engelmi replied:
Good point. Added a `node_is_online` check here.

alexlarsson commented on src/proxy/main.c:
```diff
@@ -0,0 +1,118 @@
+static void usage(char *argv[]) {
+        printf("Usage: %s \n"
+               "\t[-h help]\t\t Print this help message.\n"
+               "\t<node>_<service>\t Input the name of the node the required service is supposed to run on.\n",
```
I guess this should have two underscores between node and service.

alexlarsson replied:
Or alternatively we could use just one. It seems like systemd normaly uses dashes to separate words, like "abrt-journal-core.service", and a single dash won't conflict with that.

alexlarsson replied:
Only exceptions on my machine are:

- proc-sys-fs-binfmt_misc.automount
- proc-sys-fs-binfmt_misc.mount,
- var-lib-nfs-rpc_pipefs.mount
- wpa_supplicant.service
- rpc_pipefs.target

So, probably fine.

alexlarsson replied:
I changed this to use a single dash

alexlarsson commented on src/proxy/main.c:
```diff
@@ -0,0 +1,118 @@
+        _cleanup_free_ char *token = NULL;
+        token = strtok_r(rest, "__", &rest);
+        *ret_node_name = strdup(token);
+        *ret_unit_name = strdup(strtok_r(rest, "__", &rest));
```
Wold not hurt to do some validation of the format here, to ensure we're not returning a NULL unit name

alexlarsson commented on src/proxy/main.c:
```diff
@@ -0,0 +1,118 @@
+                        unit_name,
+                        "replace");
+        if (r < 0) {
+                hirte_log_errorf("Failed to create proxy: %s", strerror(-r));
```
I don't think it is the right thing to log an error here. This will return an error if the remote service failed to start for whatever reason. I mean, this is an error in some sense, but it is still expected and natural for the proxy service that this happens.

Maybe we can report a specific error code for that case and just log an error for other error codes.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+        } else if (active_state == UNIT_FAILED) {
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_errorf(
+                                proxy->request_message, SD_BUS_ERROR_FAILED, "Proxy service failed to start");
```
So, here, lets send a custom error message code, like org.containers.hirte.Error.RemoteUnitFailed, then we can handle it in the proxy binary.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
```
I think we should add some verification here that the reply to the initial request is never sent twice. For example:
```
if (proxy->request_message != NULL) {
  ....
  sd_bus_message_unref(proxy->request_message);
  proxy->request_message = NULL;
}
```

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+        proxy->id = ++next_id;
+        proxy->agent = agent_ref(agent);
+        proxy->request_message = sd_bus_message_ref(request_message);
+        LIST_INIT(proxy_jobs, proxy);
```
Eventually we need the agent to Stop the proxy service, so we need the actual service name of the proxy service. So, we should probably change the proxy to just send the entire service name to the agent and do the parsing of its format in the agent instead.

alexlarsson replied:
Alternatively we could have the proxy binary send both the split up parts and the service name. Unclear which is best.

alexlarsson replied:
I changed this to pass the local service name. This makes the dbus api very flexible, and all the "weirdness" is in the proxy and the template.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+
+        proxy->ref_count = 1;
+        proxy->id = ++next_id;
+        proxy->agent = agent_ref(agent);
```
This refs the agent, which is a hard ref, but proxy.h comment says this is a weak ref, so either of these are wrong.

I think it needs to be a weak ref actually, because otherwise there is a refcount cycle. I.e. as long as you create any kind of proxy it will keep the agent alive. I.e. you can't unref your initial refcount of the agent and have it destroy any outstanding proxies.

Obviously in the specific case of the agent object this will never be unref:ed in the lifetime of the program, but it is still nice to get these things right. What you want is:

* Main owns the main agent ref.
* Agent owns the proxies via the list (i.e. if you remove from list you unref)
* Proxy has a weak ref (i.e. pointer, but not ref:ed) back to the agent

This means in theory the proxy can outlive its agent, so when removing the proxy from the agent we should probably NULL out ->agent, etc.

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+                return;
+        }
+
+        LIST_REMOVE(proxy_jobs, proxy->agent->outstanding_proxy_jobs, proxy);
```
This is not quite right LIST_REMOVE assumes the node is already in the list. This is not always the case at the moment. agent_method_create_proxy() adds it to the list, but not before doing some operations that can fail. If those fail the unref will try to remove the proxy even though its not in the list.

A better approach is to always handle the list add/remove in the same place, i.e. the agent. This means, that we remove the proxy from the list (and then unref it) in agent_method_remove_proxy()  and the final unref.

alexlarsson replied:
This is related to the above weak ref comment.

alexlarsson replied:
Additionally, this means we probably need to have an proxy_service_unexport() too, which we call when removing the proxy from the agent list.

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1038,6 +1045,81 @@ static const sd_bus_vtable internal_agent_vtable[] = {
+        r = proxy_job_emit_proxy_new(proxy);
+        if (r < 0) {
+                hirte_log_errorf("Failed to emit ProxyNew signal: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
```
This isn't right I think. We should either report an error to the, and then *not* add the job to the list, or alternatively, not report an error if we added the proxy to the list.

I think the later is maybe the correct approach. What if the agent is not yet connected at the moment, but later connects with the manager. We then re-send the ProxyNew, and will eventually get a reply.

alexlarsson replied:
Actually maybe we should do both. Fail the call if any random error, but silently succeed if emit_proxy_new() returns the specific error for "not online yet".

alexlarsson replied:
(Alternatively do the is-online check before calling emit)

alexlarsson commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,135 @@
+}
+
+
+int proxy_job_emit_proxy_new(ProxyJob *proxy) {
```
These emit functions should check that `agent->connection_state == AGENT_CONNECTION_STATE_CONNECTED`, or otherwise return some error. Otherwise we might try to send a message on peer_dbus during the connection cycle.

As I said above, if we are disconnected when this is called, we will call it again when connecting.

alexlarsson commented on src/manager/node.c:
```diff
@@ -303,9 +304,10 @@ static int node_match_unit_state_changed(sd_bus_message *m, void *userdata, UNUS
         UnitSubscription *usub = NULL;
-        LIST_FOREACH(subs, usub, usubs->subs) {
+        UnitSubscription *next_usub = NULL;
+        LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
```
If we change to FOREACH_SAFE here I think it is probably best to do it in all the LIST_FOREACH emits too.

alexlarsson commented on src/manager/node.c:
```diff
@@ -377,6 +378,55 @@ static int node_match_heartbeat(UNUSED sd_bus_message *m, UNUSED void *userdata,
+        Manager *manager = requesting_node->manager;
+
+        Node *providing_node = manager_find_node(manager, node_name);
+        if (providing_node == NULL) {
```
In this, and the node-is-offline case i think it would make more sense to report a specific error code and error message (via the sd_bus_error argument). That should make the proper reason for the failure appear in the journal log on the node, which will help with debugging.

alexlarsson commented on src/manager/node.c:
```diff
@@ -377,6 +378,55 @@ static int node_match_heartbeat(UNUSED sd_bus_message *m, UNUSED void *userdata,
+                hirte_log_error("Failed to create proxy monitor, OOM");
+                return -ENOMEM;
+        }
+        LIST_APPEND(monitors, manager->proxy_monitors, proxy_monitor_ref(monitor));
```
I don't think we should add the monitor to the list untill we know everything is successful (i.e. the subscription proxy new call succeeds. Otherwise we'll get a half-working proxy in the list which will never complete.

alexlarsson commented on src/manager/node.c:
```diff
@@ -433,6 +434,46 @@ static int node_on_match_proxy_new(sd_bus_message *m, void *userdata, UNUSED sd_
+        (void) sd_bus_message_rewind(m, true);
+        (void) sd_bus_message_skip(m, "s"); /* skip node name */
+
+        return node_run_unit_lifecycle_method(m, providing_node, "stop", "StopUnit");
```
This is not at all right. If some proxy on some node no longer depends on a remove service that doesn't mean we should stop it. Some other proxy, or a local service on the node may very well still depend on it.  For now, all we need to do is clean up the proxy monitor in the manager because we no longer need to track the remote service being stopped.

There has been discussions of how we can handle remote stop-dependencies, but this is a very tricky area, rife with race conditions and semantical details. Lets just ignore it for now.

alexlarsson commented on src/manager/node.c:
```diff
@@ -433,6 +434,46 @@ static int node_on_match_proxy_new(sd_bus_message *m, void *userdata, UNUSED sd_
+                return -ENXIO;
+        }
+        if (!node_is_online(providing_node)) {
+                hirte_log_errorf("Node '%s' is offline", node_name);
```
We should be able to stop the monitor even if the remove node is offline.

alexlarsson commented on src/manager/node.c:
```diff
@@ -724,6 +724,20 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
+                            streq(proxy_monitor->requesting_node->name, node->name)) {
+                                proxy_monitor_close(proxy_monitor);
+                                manager_remove_proxy_monitor(manager, proxy_monitor);
+                        }
```
I don't think this is the right way to handle this. We already sent the (virtual) unit events, and I think our handling of the termination of the proxy should be the same, independent of whether the events were virtual or real.

In other words, if we get *actual* reported events from the node that the service was stopped or unloaded, then we should report that to the requesting node and then stop the monitor. That could should also trigger from the virtual events, and thus there is no need for this special cleanup code.

alexlarsson replied:
I guess we still want to do this if the requesting node disconnects though.

alexlarsson replied:
Although if we drop the proxy request when there requesting agent disconnects like this, we need the agent to realize that if the manager disconnects, then it needs to stop all outstanding proxies (both the initial call, and send a stop to all active proxies).

alexlarsson commented on src/manager/monitor_proxy.c:
```diff
@@ -0,0 +1,262 @@
+                r = proxy_monitor_send_state(monitor, state);
+                proxy_monitor_close(monitor);
+                manager_remove_proxy_monitor(monitor->manager, monitor);
+        }
```
It is a bit weird that we track failure to activate in two places, both in the proxy monitor, as well as in the Job callback (proxy_start_unit_callback). Its quite possible that you report the same thing twice, or that something else causes a state_change before the actual job completes. 

I think the proper way to handle this is to start the job on new proxy request, and only reply to that request with the resulting status from the job request. The monitor should only be uses to track the state of the unit after the job is finished. Then when that reports the unit as stopped we send a message to the agent to stop the proxy.

alexlarsson replied:
Hmm, that would require not only calling StartUnit, but also tracking the job that results from that.
Also, would need the job monitoring to be robust against agent disconnect. So, maybe using the monitor to track start is a better approach anyway?

But if so, I wonder if the current approach is robust enough. It does nothing until the state changes to active or failed, but maybe there are other possible scenarios. Maybe it was it "deactivating" state when the monitor was created, and then it goes into inactive state. Although in that case the StartJob should probably trigger a new start. I dunno, it feels like there are potential problems here.

alexlarsson replied:
Anyway, no matter what, we need to be more careful about whether we reported the new state for the ProxyMonitor, be it from on_unit_state_changed, or proxy_start_unit_callback, because we need to guarantee that we never respond twice independent of whatever order things happens in.

alexlarsson committed:
manager: Make the monitor object usable for in-memory monitors too

We want to user monitors for proxies, so they need to be useful
without being dbus objects.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
Support proxies in the manager

Implements proxy monitor to start and monitor requested proxy service
units as well as update the state of requesting agent

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
agent: Clean up logging wrt to (re)connection

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
agent: Fix reporting of unit state changes

We initialize the unit state for new unit infos to inactive(dead) which
is the default state for systemd services. This way we don't report
an initial change to dead whenever some property changes.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

Yarboa commented on src/agent/agent.c:
```diff
@@ -1639,7 +1927,7 @@ static bool agent_connect(Agent *agent) {
         }
 
-        hirte_log_infof("Agent '%s' connection attempt %d: success", agent->name, agent->connection_retry_count);
+        hirte_log_infof("Connected to manager as '%s'", agent->name);
```
You should change the grep in e2e 
https://github.com/containers/hirte/blob/main/.github/workflows/e2e.yml#L71-L73
https://github.com/containers/hirte/blob/main/.github/workflows/e2e.yml#L88-L90

Consider changing from 
 grep \\"Agent 'bar' connection attempt 0: success\\"" -> grep \\"Connected to manager as  'bar'\\""

engelmi commented on doc/dbus-interfaces.md:
```diff
@@ -297,7 +317,8 @@ This is the main interface that the node implements and that is used by the mana
 
   * `ProxyRemoved(s nodeName, s unitName, o proxy)`
```
```suggestion
  * `ProxyRemoved(s nodeName, s unitName)`
```

engelmi commented on doc/man/hirte-proxy.1.md:
```diff
@@ -0,0 +1,53 @@
+local D-Bus to request services required by a local unit to be started
+or stopped on a remote node managed by another `hirte-agent`.
+
+**hirte [OPTIONS]**
```
```suggestion
**hirte-proxy [OPTIONS]**
```

engelmi commented on src/proxy/main.c:
```diff
@@ -0,0 +1,185 @@
+        opt_operation = argv[optind];
+        opt_node_unit = argv[optind + 1];
+
+        if (!streqi(opt_operation, "create") && !streqi(opt_operation, "remove")) {
```
```suggestion
        if (!streqi(opt_operation, OPT_OPERATION_CREATE) && !streqi(opt_operation, OPT_OPERATION_REMOVE)) {
```

engelmi commented on src/proxy/main.c:
```diff
@@ -0,0 +1,185 @@
+                r = call_proxy_method(api_bus, "CreateProxy", opt_node_unit, node_name, unit_name);
+        } else if (streqi(opt_operation, "remove")) {
+                r = call_proxy_method(api_bus, "RemoveProxy", opt_node_unit, node_name, unit_name);
+        }
```
```suggestion
        if (streqi(opt_operation, OPT_OPERATION_CREATE)) {
                r = call_proxy_method(api_bus, "CreateProxy", opt_node_unit, node_name, unit_name);
        } else if (streqi(opt_operation, OPT_OPERATION_REMOVE)) {
                r = call_proxy_method(api_bus, "RemoveProxy", opt_node_unit, node_name, unit_name);
        }
```

engelmi commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,327 @@
+}
+
+bool proxy_service_export(ProxyService *proxy) {
+        int r = sd_bus_add_object_vtable(
```
I think the `data/org.containers.hirte.internal.Proxy.xml` needs to be updated accordingly. Currently, it only contains a `Ready` method.

engelmi commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,327 @@
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        hirte_log_errorf("Got proxy new: %s", reason);
```
I think this should be an info log as nothing went wrong.

engelmi commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,327 @@
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        hirte_log_errorf("Got proxy removed: %s", reason);
```
I think this should be an info log as nothing went wrong.

engelmi commented on src/agent/proxy.c:
```diff
@@ -0,0 +1,327 @@
+        }
+
+        hirte_log_errorf("Got proxy new: %s", reason);
+
```
Do we need to `proxy_service_initial_state_reached(proxy, false);` here as well?

alexlarsson replied:
No, we always get a new event when something is starting to activate. Basically when systemd instantiates the unit in memory. We don't yet know anything about how it activates.

engelmi commented on src/agent/agent.c:
```diff
@@ -1011,6 +1070,103 @@ static int agent_method_unsubscribe(sd_bus_message *m, void *userdata, UNUSED sd
+ ************************************************************************/
+
+static char *get_dep_unit(const char *unit_name) {
+        return strcat_dup("hirte-dep@", unit_name);
```
Could we move `hirte-dep@` to a constant, e.g. in protocol.h (although I am not sure if it makes sense to move it there)?

alexlarsson replied:
I dunno, its already abstracted away into a single function. Not sure having it in the header helps.

engelmi replied:
Good point. Lets keep it like it is.

engelmi commented on src/agent/agent.c:
```diff
@@ -1039,6 +1195,120 @@ static const sd_bus_vtable internal_agent_vtable[] = {
                         0),
         SD_BUS_SIGNAL_WITH_NAMES("UnitRemoved", "s", SD_BUS_PARAM(unit), 0),
         SD_BUS_SIGNAL_WITH_NAMES("Heartbeat", "s", SD_BUS_PARAM(agent_name), 0),
+        SD_BUS_METHOD("StartDep", "s", "", agent_method_start_dep, 0),
```
I think this is missing in `data/org.containers.hirte.internal.Agent.xml`

alexlarsson committed:
Add doc/proxies.md describing some details about proxies.

Signed-off-by: Alexander Larsson <alexl@redhat.com>
